File: YGHorizontalView.m
//
//  YGHorizontalView.m
//  AllYoga
//
//  Created by liwei on 2017/7/18.
//  Copyright © 2017年 biyunkeji. All rights reserved.
//

#import "YGHorizontalView.h"

@interface YGHorizontalView ()
/** 左边文字 */
@property (nonatomic,strong) UILabel *frontLabel;

/** 中间图片 */
@property (nonatomic,strong) UIImageView *middleImageView;

/**  右边文字 */
@property (nonatomic,strong) UILabel *trailLabel;

@end

@implementation YGHorizontalView

-(UILabel *)trailLabel{
    if (!_trailLabel){
        _trailLabel = [[UILabel alloc] init];
        _trailLabel.textAlignment = self.trailAlign;
        _trailLabel.font = self.trailFont;
        _trailLabel.textColor = self.trailColor;
        [self addSubview:_trailLabel];
    }
    return _trailLabel;
}

-(UILabel *)frontLabel{
    if (!_frontLabel){
        _frontLabel = [[UILabel alloc] init];
        _frontLabel.textColor = self.frontColor;
        _frontLabel.font = self.frontFont;
        _frontLabel.textAlignment = self.frontAlign;
        [self addSubview:_frontLabel];
    }
    return _frontLabel;
}

-(UIImageView *)middleImageView{
    if (!_middleImageView){
        _middleImageView = [[UIImageView alloc] init];
        _middleImageView.contentMode = self.contentMode;
        [self addSubview:_middleImageView];
    }
    return _middleImageView;
}

- (instancetype)initWithFrame:(CGRect)frame {

    if (self = [super initWithFrame:frame]) {
        [self initProperties];
    }
    return  self;
}

- (instancetype)init {

    if (self = [super init]) {
        [self initProperties];
    }
    return self;
}

- (void)initProperties {

    self.frontFont = [UIFont systemFontOfSize:15];
    self.frontAlign = NSTextAlignmentLeft;
    self.frontColor = [UIColor blackColor];
    
    self.trailFont = [UIFont systemFontOfSize:15];
    self.trailColor = [UIColor blackColor];
    self.trailAlign = NSTextAlignmentLeft;
    
    self.contentMode = UIViewContentModeScaleAspectFit;
    
    self.layoutDirection = YGHorizontalLayoutDirectionLeft;
    
    self.imageViewInset = UIEdgeInsetsMake(5, 10, 5, 10);
    
    self.textInset = UIEdgeInsetsMake(0, 10, 0, 10);
}

- (void)setFrontContent:(NSString *)frontContent {

    _frontContent = frontContent;
    [self setNeedsLayout];
}

- (void)setTrailContent:(NSString *)trailContent {

    _trailContent = trailContent;
    [self setNeedsLayout];
}

- (void)setImageName:(NSString *)imageName {

    _imageName = imageName;
    self.middleImageView.image = [UIImage imageNamed:imageName];
    [self setNeedsLayout];
}

- (void)layoutSubviews {

    [super layoutSubviews];
    
    switch (self.layoutDirection) {
        case YGHorizontalLayoutDirectionLeft:
            [self layoutSubviewsFromFront];
            break;
        case YGHorizontalLayoutDirectionRight:
            [self layoutSubviewsFromTrail];
            break;
        default:
            [self layoutSubviewsFromMiddle];
            break;
    }
}

//从左往右布局
- (void)layoutSubviewsFromFront {

    CGFloat maxX = self.textInset.left;
    if (self.frontContent || self.frontContent.length > 0) {
        CGSize frontSize = [self.frontContent boundingRectWithSize:self.bounds.size options:NSStringDrawingUsesFontLeading attributes:@{NSFontAttributeName:self.frontFont,NSForegroundColorAttributeName:self.frontColor} context:nil].size;
        self.frontLabel.center = CGPointMake(maxX + frontSize.width * 0.5, self.frame.size.height * 0.5);
        self.frontLabel.bounds = CGRectMake(0, 0, frontSize.width, frontSize.height);
        maxX = CGRectGetMaxX(self.frontLabel.frame);
    }
    
    if (self.imageName && self.imageName.length > 0) {
        self.middleImageView.center = CGPointMake(maxX + self.imageViewInset.left + (self.frame.size.height - self.imageViewInset.top - self.imageViewInset.left) * 0.5, self.frame.size.height * 0.5);
        self.middleImageView.bounds = CGRectMake(0, 0, self.frame.size.height - self.imageViewInset.top - self.imageViewInset.left, self.frame.size.height - self.imageViewInset.top - self.imageViewInset.left);
        maxX = CGRectGetMaxX(self.middleImageView.frame);
    }
    
    if (self.trailContent || self.trailContent.length > 0) {
         CGSize trailSize = [self.trailContent boundingRectWithSize:self.bounds.size options:NSStringDrawingUsesFontLeading attributes:@{NSFontAttributeName:self.trailFont,NSForegroundColorAttributeName:self.trailColor} context:nil].size;
        self.trailLabel.center = CGPointMake(maxX + self.imageViewInset.right + trailSize.width * 0.5, self.frame.size.height * 0.5);
        self.trailLabel.bounds = CGRectMake(0, 0, trailSize.width, trailSize.height);
    }
    
}
//从中间向两边
- (void)layoutSubviewsFromMiddle {
    CGFloat maxX = self.textInset.left;
    
    if (self.imageName && self.imageName.length > 0) {
        self.middleImageView.center = CGPointMake(self.frame.size.width * 0.5, self.frame.size.height * 0.5);
        self.middleImageView.bounds = CGRectMake(0, 0, self.frame.size.height - self.imageViewInset.top - self.imageViewInset.left, self.frame.size.height - self.imageViewInset.top - self.imageViewInset.left);
    }
    
    if (self.frontContent || self.frontContent.length > 0) {
        maxX = CGRectGetMinX(self.middleImageView.frame);
        CGSize frontSize = [self.frontContent boundingRectWithSize:self.bounds.size options:NSStringDrawingUsesFontLeading attributes:@{NSFontAttributeName:self.frontFont,NSForegroundColorAttributeName:self.frontColor} context:nil].size;
        self.frontLabel.center = CGPointMake(maxX - frontSize.width * 0.5, self.frame.size.height * 0.5);
        self.frontLabel.bounds = CGRectMake(0, 0, frontSize.width, frontSize.height);
    }
    
    if (self.trailContent || self.trailContent.length > 0) {
        maxX = CGRectGetMaxX(self.middleImageView.frame);
        CGSize trailSize = [self.trailContent boundingRectWithSize:self.bounds.size options:NSStringDrawingUsesFontLeading attributes:@{NSFontAttributeName:self.trailFont,NSForegroundColorAttributeName:self.trailColor} context:nil].size;
        self.trailLabel.center = CGPointMake(maxX + self.imageViewInset.right + trailSize.width * 0.5, self.frame.size.height * 0.5);
        self.trailLabel.bounds = CGRectMake(0, 0, trailSize.width, trailSize.height);
    }
}
//从右边向左边
- (void)layoutSubviewsFromTrail {
    CGFloat minX = self.frame.size.width - self.textInset.right;
    if (self.trailContent || self.trailContent.length > 0) {
        CGSize trailSize = [self.trailContent boundingRectWithSize:self.bounds.size options:NSStringDrawingUsesFontLeading attributes:@{NSFontAttributeName:self.trailFont,NSForegroundColorAttributeName:self.trailColor} context:nil].size;
        self.trailLabel.center = CGPointMake(minX - trailSize.width * 0.5, self.frame.size.height * 0.5);
        self.trailLabel.bounds = CGRectMake(0, 0, trailSize.width, trailSize.height);
        minX = CGRectGetMinX(self.frontLabel.frame);
    }
    
    if (self.imageName && self.imageName.length > 0) {
        self.middleImageView.center = CGPointMake(minX - self.imageViewInset.right - (self.frame.size.height - self.imageViewInset.top - self.imageViewInset.left) * 0.5, self.frame.size.height * 0.5);
        self.middleImageView.bounds = CGRectMake(0, 0, self.frame.size.height - self.imageViewInset.top - self.imageViewInset.left, self.frame.size.height - self.imageViewInset.top - self.imageViewInset.left);
        minX = CGRectGetMinX(self.middleImageView.frame);
    }
    
    if (self.frontContent || self.frontContent.length > 0) {
        CGSize frontSize = [self.frontContent boundingRectWithSize:self.bounds.size options:NSStringDrawingUsesFontLeading attributes:@{NSFontAttributeName:self.frontFont,NSForegroundColorAttributeName:self.frontColor} context:nil].size;
        self.trailLabel.center = CGPointMake(minX - self.imageViewInset.left - frontSize.width * 0.5, self.frame.size.height * 0.5);
        self.trailLabel.bounds = CGRectMake(0, 0, frontSize.width, frontSize.height);
    }

}

@end

File: test/test/YGHorizontalView.swift
//
//  YGHorizontalView.swift
//  AllYoga-swift
//
//  Created by liwei on 2017/6/27.
//  Copyright © 2017年 liwei. All rights reserved.
//

import UIKit

class YGHorizontalView: UIControl {

    enum LayoutDirection {
        case left//从左往右布局
        case right//从右往左布局
        case center//从中间向两边布局
    }
    
    //默认从左往右布局
    var layoutDirection:LayoutDirection = .left
    //图片名称
    var imageName:String? {
    
        didSet {
            contentImageView.image = UIImage(named: imageName!)
            setNeedsLayout()
        }
    }
    //前半段文字
    var beforeHalfText:String? {
    
        didSet {
            beforeHalfLabel.text = beforeHalfText
            setNeedsLayout()
        }
    }
    //后半段文字
    var behindHalfText:String? {
    
        didSet {
            behindHalfLabel.text = behindHalfText
            setNeedsLayout()
        }
    }
    //前半段文字颜色
    var beforeHalfColor:UIColor? = UIColor.black
    //后半段文字颜色
    var behindHalfColor:UIColor? = UIColor.black
    //前半段文字大小
    var beforeHalfFont:UIFont? = UIFont.systemFont(ofSize: 15)
    //后半段文字大小
    var behindHalfFont:UIFont? = UIFont.systemFont(ofSize: 15)
    
    
    //前半段文字的align
    var beforeHalfAlign:NSTextAlignment = .left
    //后半段文字的align
    var behindHalfAlign:NSTextAlignment = .right
    //图片的填充模式
    var imageMode:UIViewContentMode = .scaleAspectFit
    //控件之间的间距，默认是5
    var padding:CGFloat = 5
    //图片高度
    var imageHeight:CGFloat  = 0
    
    
   private lazy var contentImageView:UIImageView! = {
    
        let image = UIImageView()
        image.contentMode = self.imageMode
        self.addSubview(image)
        return image
    }()
    
   private lazy var beforeHalfLabel:UILabel! = {
    
        let label = UILabel()
        label.textAlignment = self.beforeHalfAlign
        label.textColor = self.beforeHalfColor
        label.font = self.beforeHalfFont
        self.addSubview(label)
        return label
    }()
    
   private lazy var behindHalfLabel:UILabel! = {
        
        let label = UILabel()
        label.textAlignment = self.behindHalfAlign
        label.textColor = self.behindHalfColor
        label.font = self.behindHalfFont
        self.addSubview(label)
        return label
    }()
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        switch layoutDirection {
        case .left:
            layoutFromLeft()
            break
        case .right:
            layoutFromRight()
            break;
        default:
            layoutFromCenter()
            break;
        }
        
    }
    
    private func layoutFromLeft() {
    
    
        let maxSize = frame.size
        var maxX = padding
        
        if let beforeText = beforeHalfText {
            let beforeSize = stringSize(string: beforeText, maxSize: maxSize, font: beforeHalfFont!, color: beforeHalfColor)
            beforeHalfLabel.center = CGPoint(x: maxX + beforeSize.width * 0.5, y: frame.height * 0.5)
            beforeHalfLabel.bounds = CGRect(x: 0, y: 0, width: beforeSize.width, height: beforeSize.height)
            maxX = beforeHalfLabel.frame.maxX + padding
        }
        
        if imageName != nil {
            let imageW =  imageHeight == 0 ? frame.height - 2 * 10 : imageHeight
            
           contentImageView.center = CGPoint(x: maxX + imageW * 0.5, y: frame.height * 0.5)
            contentImageView.bounds = CGRect(x: 0, y: 0, width: imageW, height: imageW)
            maxX = contentImageView.frame.maxX + padding
        }
        
        if let behindText = behindHalfText {
            let beforeSize = stringSize(string: behindText, maxSize: maxSize, font: behindHalfFont!, color: behindHalfColor)
            behindHalfLabel.center = CGPoint(x: maxX + beforeSize.width * 0.5, y: frame.height * 0.5)
            behindHalfLabel.bounds = CGRect(x: 0, y: 0, width: beforeSize.width, height: beforeSize.height)
        }
    }
    
    private func layoutFromRight() {
        let maxSize = frame.size
        var minX = padding
        
        if let behindText = behindHalfText {
            let beforeSize = stringSize(string: behindText, maxSize: maxSize, font: behindHalfFont!, color: behindHalfColor)
            behindHalfLabel.center = CGPoint(x: frame.width - minX - beforeSize.width * 0.5, y: frame.height * 0.5)
            behindHalfLabel.bounds = CGRect(x: 0, y: 0, width: beforeSize.width, height: beforeSize.height)
            minX = behindHalfLabel.frame.minX - padding
        }

        if imageName != nil {
            let imageW =  imageHeight == 0 ? frame.height - 2 * 10 : imageHeight
            
            contentImageView.center = CGPoint(x: minX - imageW * 0.5, y: frame.height * 0.5)
            contentImageView.bounds = CGRect(x: 0, y: 0, width: imageW, height: imageW)
            minX = contentImageView.frame.minX - padding
        }

        if let beforeText = beforeHalfText {
            let beforeSize = stringSize(string: beforeText, maxSize: maxSize, font: beforeHalfFont!, color: beforeHalfColor)
            beforeHalfLabel.center = CGPoint(x: minX - beforeSize.width * 0.5, y: frame.height * 0.5)
            beforeHalfLabel.bounds = CGRect(x: 0, y: 0, width: beforeSize.width, height: beforeSize.height)
        }
        
   }
    
    private func layoutFromCenter() {
        let maxSize = frame.size
        if imageName != nil {
            let imageW =  imageHeight == 0 ? frame.height - 2 * 10 : imageHeight
            
            contentImageView.center = CGPoint(x: frame.width * 0.5, y: frame.height * 0.5)
            contentImageView.bounds = CGRect(x: 0, y: 0, width: imageW, height: imageW)
        }
        
        
        if let beforeText = beforeHalfText {
            let minX = imageName == nil ? frame.width * 0.5 : contentImageView.frame.minX
            
            let beforeSize = stringSize(string: beforeText, maxSize: maxSize, font: beforeHalfFont!, color: beforeHalfColor)
            beforeHalfLabel.center = CGPoint(x: minX - padding - beforeSize.width * 0.5, y: frame.height * 0.5)
            beforeHalfLabel.bounds = CGRect(x: 0, y: 0, width: beforeSize.width, height: beforeSize.height)
        }

        
        if let behindText = behindHalfText {
            let maxX = imageName == nil ? frame.width * 0.5 : contentImageView.frame.maxX

            let beforeSize = stringSize(string: behindText, maxSize: maxSize, font: behindHalfFont!, color: behindHalfColor)
            behindHalfLabel.center = CGPoint(x: maxX + padding + beforeSize.width * 0.5, y: frame.height * 0.5)
            behindHalfLabel.bounds = CGRect(x: 0, y: 0, width: beforeSize.width, height: beforeSize.height)
        }
        
    }
}

extension YGHorizontalView {

    public func stringLength(string:String?) -> Int {
        if string == nil {
            return 0
        }
        return string!.characters.count
    }
    
    public func stringSize(string:String?,maxSize:CGSize,font:UIFont, color:UIColor?) -> CGSize {
        var textColor:UIColor!
        
        if color == nil {
            textColor = UIColor.white
        }else {
            textColor = color!
        }
        var content:String = ""
        if  self.stringLength(string:string) > 0  {
            content = string!
        }
        let size = (content as NSString).boundingRect(with: maxSize, options: .usesLineFragmentOrigin, attributes: [NSFontAttributeName:font,NSBackgroundColorAttributeName:textColor], context: nil).size
        
        return size
    }

}
